robotiq_joint_state_publisher/src/s_model_joint_states.cpp

```diff
@@ -0,0 +1,208 @@
+  Finger finger_right;									///< Robotiq FINGER B
+  Finger finger_middle;									///< Robotiq FINGER C
+  int scissor;										///< Scissor position
+  int mode;										///< Robotiq grasping mode
+  std::vector<double> joint_positions;							///< All the joint values
+};
+
+/**
+ * Calculate scissor_joint based on the scissor position.
+ */
+inline double Robotiq3::scissor_joint() const {
+  // fully open is +20, fully closed is -10
+  // std::cout << scissor << std::endl;
+  if(0 <= scissor && scissor <= 15) return 20.0*DEG_TO_RAD;
+  else if(15 < scissor && scissor <= 240) return (20.0-30.0*((scissor-15.0)/225.0))*DEG_TO_RAD;
+  else return -10.0*DEG_TO_RAD;
+}
+
+/**
+ * Callback function for "SModelRobotInput" topic.
+ * */
+void Robotiq3::callback(const robotiq_s_model_control::SModel_robot_input::ConstPtr &msg) {
+  finger_left = Finger(msg->gPOA);		// set left finger position
+  finger_right = Finger(msg->gPOB);		// set right finger position
+  finger_middle = Finger(msg->gPOC);		// set middle finger position
+  scissor = msg->gPOS;				// set scissor position
+  mode = msg->gMOD;				// set mode
+
+  // if (mode == BASIC)        scissor = 137;
+  // else if (mode == PINCH)   scissor = 255;
+  // else if (mode == WIDE)    scissor = 0;
+  // else if (mode == SCISSOR) scissor = 0;
+
+  // Set all the joint values
+  joint_positions.at(0)  =  scissor_joint();
+  joint_positions.at(1)  =  finger_left.joint_1();
+  joint_positions.at(2)  =  finger_left.joint_2();
+  joint_positions.at(3)  =  finger_left.joint_3();
+  joint_positions.at(4)  = -joint_positions.at(0);
+  joint_positions.at(5)  =  finger_right.joint_1();
+  joint_positions.at(6)  =  finger_right.joint_2();
+  joint_positions.at(7)  =  finger_right.joint_3();
+  joint_positions.at(8)  =  finger_middle.joint_1();
+  joint_positions.at(9)  =  finger_middle.joint_2();
+  joint_positions.at(10) =  finger_middle.joint_3();
+}
+
+/**
+ * Main method.
+ */
+int main(int argc, char *argv[]) {
+
+  // Create Robotiq3
+  Robotiq3 robotiq;
+
+  // ROS init, nodehandle, and rate
+  ros::init(argc, argv, "s_model_joint_states");
+  ros::NodeHandle nh;
+  ros::Rate loop_rate(20);  // Hz
+
+  // joint state publisher
+  ros::Publisher joint_pub;
+  joint_pub = nh.advertise<sensor_msgs::JointState>("joint_states", 10);
+
+  // robotiq state message subscriber
+  ros::Subscriber joint_sub;
+  joint_sub = nh.subscribe("SModelRobotInput", 10, &Robotiq3::callback, &robotiq);
+
+  // wait until the URDF has been loaded to avoid publishing joints prematurely
+  while (!ros::isShuttingDown() && !nh.hasParam("/robot_description")) {
+    ROS_ERROR("[s_model_joint_states] Waiting for URDF");
+    ros::Duration(0.1).sleep();
+  }
+
+  // set pseudo-namespace
+  std::string gripper_prefix;
+  if (argc > 1) gripper_prefix = argv[1];
+  else gripper_prefix = "";
+
+  // joint names for sensor_msgs::JointState message
+  // order matters!
+  std::vector<std::string> joint_names(11, "");
```

Comment:
The `joint_names` and `joint_positions` vectors split across the node and the `Robotiq3` class, but they are required to be the same length.  Do you think it would be better to embed the joint names in the `Robotiq3` class?

I'd also recommend using and enumeration for the name/position indices(old school) or perhaps a `std::map`